From ca3cc052a06e83cde0c28a8c73e24a837ef5fb79 Mon Sep 17 00:00:00 2001
From: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
Date: Mon, 27 Feb 2017 16:17:57 +0300
Subject: [PATCH 14/17] mic: vop: avoid extra rescheduling on each interrupt

In common case, VOP host-side interrupt hanfler only needs to signal
a wakeup. No need to enter workqueue context for that.

On heavy network usage, this change removes tons if reschedules.

Signed-off-by: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
---
 drivers/misc/mic/vop/vop_vringh.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/drivers/misc/mic/vop/vop_vringh.c b/drivers/misc/mic/vop/vop_vringh.c
index 927d71d9..2039137 100644
--- a/drivers/misc/mic/vop/vop_vringh.c
+++ b/drivers/misc/mic/vop/vop_vringh.c
@@ -158,7 +158,15 @@ static irqreturn_t _vop_virtio_intr_handler(int irq, void *data)
 	struct vop_device *vpdev = vdev->vpdev;
 
 	vpdev->hw_ops->ack_interrupt(vpdev, vdev->virtio_db);
-	schedule_work(&vdev->virtio_bh_work);
+
+	if (unlikely(vdev->dc->used_address_updated) ||
+	    unlikely(vdev->dc->vdev_reset))
+		schedule_work(&vdev->virtio_bh_work);
+	else {
+		vdev->poll_wake = 1;
+		wake_up(&vdev->waitq);
+	}
+
 	return IRQ_HANDLED;
 }
 
-- 
1.9.1

